File: problem/Power_Grid_Maintenance/processQueries.c
#include <stdlib.h>
#include <string.h>

typedef struct
{
    int *data;
    int size;
    int cap;
} MinHeap;

static void heap_init(MinHeap *h, int cap)
{
    h->data = (int *)malloc(sizeof(int) * (cap > 0 ? cap : 1));
    h->size = 0;
    h->cap = (cap > 0 ? cap : 1);
}
static void heap_free(MinHeap *h)
{
    free(h->data);
}
static void heap_ensure(MinHeap *h)
{
    if (h->size == h->cap)
    {
        h->cap = h->cap ? h->cap * 2 : 4;
        h->data = (int *)realloc(h->data, sizeof(int) * h->cap);
    }
}
static void heap_push(MinHeap *h, int x)
{
    heap_ensure(h);
    int i = h->size++;
    h->data[i] = x;
    while (i)
    {
        int p = (i - 1) >> 1;
        if (h->data[p] <= h->data[i])
            break;
        int t = h->data[p];
        h->data[p] = h->data[i];
        h->data[i] = t;
        i = p;
    }
}
static int heap_top(MinHeap *h) { return h->data[0]; }
static void heap_pop(MinHeap *h)
{
    if (h->size == 0)
        return;
    int n = --h->size;
    if (n == 0)
        return;
    h->data[0] = h->data[n];
    int i = 0;
    while (1)
    {
        int l = i * 2 + 1, r = l + 1, b = i;
        if (l < h->size && h->data[l] < h->data[b])
            b = l;
        if (r < h->size && h->data[r] < h->data[b])
            b = r;
        if (b == i)
            break;
        int t = h->data[i];
        h->data[i] = h->data[b];
        h->data[b] = t;
        i = b;
    }
}

typedef struct
{
    MinHeap *heaps;
    int heapsN;
    int *parent;
    int *sz;
    int *offline;
    int n;
} DSUEnv;

static void env_init(DSUEnv *e, int c)
{
    e->n = c;
    e->parent = (int *)malloc(sizeof(int) * c);
    e->sz = (int *)malloc(sizeof(int) * c);
    e->offline = (int *)calloc(c, sizeof(int));
    e->heaps = (MinHeap *)malloc(sizeof(MinHeap) * c);
    e->heapsN = c;
    for (int i = 0; i < c; i++)
    {
        e->parent[i] = i;
        e->sz[i] = 1;
        heap_init(&e->heaps[i], 1);
        heap_push(&e->heaps[i], i);
    }
}
static void env_free(DSUEnv *e)
{
    for (int i = 0; i < e->heapsN; i++)
        heap_free(&e->heaps[i]);
    free(e->heaps);
    free(e->parent);
    free(e->sz);
    free(e->offline);
}
static int findp(DSUEnv *e, int x)
{
    if (e->parent[x] == x)
        return x;
    e->parent[x] = findp(e, e->parent[x]);
    return e->parent[x];
}
static void merge_heaps(MinHeap *to, MinHeap *from)
{
    for (int i = 0; i < from->size; i++)
        heap_push(to, from->data[i]);
    from->size = 0;
}
static void dsu_union(DSUEnv *e, int u, int v)
{
    int up = findp(e, u);
    int vp = findp(e, v);
    if (up == vp)
        return;
    if (e->sz[up] > e->sz[vp])
    {
        e->sz[up] += e->sz[vp];
        e->parent[vp] = up;
        merge_heaps(&e->heaps[up], &e->heaps[vp]);
    }
    else
    {
        e->sz[vp] += e->sz[up];
        e->parent[up] = vp;
        merge_heaps(&e->heaps[vp], &e->heaps[up]);
    }
}

int *processQueries(int c, int **connections, int connectionsSize, int *connectionsColSize, int **queries, int queriesSize, int *queriesColSize, int *returnSize)
{
    DSUEnv E;
    env_init(&E, c);

    for (int i = 0; i < connectionsSize; i++)
    {
        int u = connections[i][0] - 1;
        int v = connections[i][1] - 1;
        dsu_union(&E, u, v);
    }

    int *ans = (int *)malloc(sizeof(int) * queriesSize);
    int an = 0;

    for (int i = 0; i < queriesSize; i++)
    {
        int type = queries[i][0];
        int node = queries[i][1] - 1;
        if (type == 1)
        {
            if (E.offline[node] == 0)
            {
                ans[an++] = node + 1;
                continue;
            }
            int p = findp(&E, node);
            while (E.heaps[p].size > 0 && E.offline[heap_top(&E.heaps[p])])
                heap_pop(&E.heaps[p]);
            if (E.heaps[p].size == 0)
                ans[an++] = -1;
            else
                ans[an++] = heap_top(&E.heaps[p]) + 1;
        }
        else
        {
            E.offline[node] = 1;
        }
    }

    *returnSize = an;
    env_free(&E);
    return ans;
}

File: problem/Power_Grid_Maintenance/processQueries.js
var processQueries = function(c, connections, queries) {
  class MinHeap {
    constructor(){ this.a = []; }
    empty(){ return this.a.length === 0; }
    top(){ return this.a[0]; }
    push(x){
      const a = this.a; a.push(x);
      let i = a.length - 1;
      while (i){
        const p = (i - 1) >> 1;
        if (a[p] <= a[i]) break;
        [a[p], a[i]] = [a[i], a[p]];
        i = p;
      }
    }
    pop(){
      const a = this.a;
      const n = a.length - 1;
      if (n < 0) return;
      if (n === 0){ a.pop(); return; }
      a[0] = a[n]; a.pop();
      let i = 0;
      while (true){
        let l = i*2 + 1, r = l + 1, b = i;
        if (l < a.length && a[l] < a[b]) b = l;
        if (r < a.length && a[r] < a[b]) b = r;
        if (b === i) break;
        [a[i], a[b]] = [a[b], a[i]];
        i = b;
      }
    }
  }

  const parent = Array.from({length: c}, (_, i) => i);
  const size = Array(c).fill(1);
  const offline = Array(c).fill(0);
  const heaps = Array.from({length: c}, () => new MinHeap());
  for (let i = 0; i < c; i++) heaps[i].push(i);

  const find = (x) => parent[x] === x ? x : (parent[x] = find(parent[x]));
  const mergeHeaps = (a, b) => { while (!b.empty()) a.push(b.top()), b.pop(); };
  const Union = (u, v) => {
    let up = find(u), vp = find(v);
    if (up === vp) return;
    if (size[up] > size[vp]) {
      size[up] += size[vp];
      parent[vp] = up;
      mergeHeaps(heaps[up], heaps[vp]);
    } else {
      size[vp] += size[up];
      parent[up] = vp;
      mergeHeaps(heaps[vp], heaps[up]);
    }
  };

  for (const [u0, v0] of connections) Union(u0 - 1, v0 - 1);

  const ans = [];
  for (const [type, node0] of queries) {
    const node = node0 - 1;
    if (type === 1) {
      if (offline[node] === 0) {
        ans.push(node + 1);
        continue;
      }
      const p = find(node);
      const h = heaps[p];
      while (!h.empty() && offline[h.top()] === 1) h.pop();
      if (h.empty()) {
        ans.push(-1);
      } else {
        ans.push(h.top() + 1);
      }
    } else {
      offline[node] = 1;
    }
  }
  return ans;
};